File: src/breadcrumbs.ts
/**
 * Breadcrumbs — automatic trail of actions before a crash.
 * Ring buffer of last 30 events: console, fetch, custom.
 * Secrets are scrubbed from messages automatically.
 *
 * Also the injection point for FullTrace's X-IW-Session-Id header — we
 * already wrap globalThis.fetch here for breadcrumb capture, so adding
 * one header to the same wrapper avoids a second monkey-patch.
 */

import type { Breadcrumb } from "./types.js"
import { injectSessionHeader } from "./fulltrace.js"

const MAX_BREADCRUMBS = 30
const breadcrumbs: Breadcrumb[] = []
let initialized = false

// Patterns that likely contain secrets
const SECRET_PATTERNS = [
  /Bearer\s+[A-Za-z0-9\-._~+\/]+=*/g,           // Bearer tokens
  /[a-zA-Z0-9_-]{20,}\.[a-zA-Z0-9_-]{20,}\.[a-zA-Z0-9_-]{20,}/g, // JWTs
  /(?:sk|pk|api|key|token|secret|password|passwd)[_-]?[:\s=]+\S{8,}/gi, // key=value secrets
  /:\/\/[^:]+:[^@]+@/g,                           // connection strings user:pass@
  /[?&](api_key|token|secret|key|password|auth|credential)=[^&\s]+/gi, // query string secrets
]

function scrubSecrets(text: string): string {
  let scrubbed = text
  for (const pattern of SECRET_PATTERNS) {
    scrubbed = scrubbed.replace(pattern, "[REDACTED]")
  }
  return scrubbed
}

/** Strip sensitive query params from URLs */
export function scrubUrl(url: string): string {
  try {
    const parsed = new URL(url, "http://localhost")
    const sensitiveParams = ["token", "key", "secret", "password", "auth", "credential", "api_key", "apiKey", "access_token"]
    for (const param of sensitiveParams) {
      if (parsed.searchParams.has(param)) {
        parsed.searchParams.set(param, "[REDACTED]")
      }
    }
    // Return just path+query if relative URL
    if (url.startsWith("/")) return parsed.pathname + parsed.search
    return parsed.href
  } catch {
    return scrubSecrets(url)
  }
}

export function addBreadcrumb(crumb: Partial<Breadcrumb> & { message: string }): void {
  breadcrumbs.push({
    timestamp: new Date().toISOString(),
    category: crumb.category ?? "custom",
    level: crumb.level ?? "info",
    message: scrubSecrets(crumb.message.slice(0, 200)),
    data: crumb.data,
  })
  if (breadcrumbs.length > MAX_BREADCRUMBS) breadcrumbs.shift()
}

export function getBreadcrumbs(): Breadcrumb[] {
  return [...breadcrumbs]
}

/**
 * Auto-intercept console and fetch to record breadcrumbs.
 * Called once from init(). Safe to call multiple times (idempotent).
 */
export function initBreadcrumbs(): void {
  if (initialized) return
  initialized = true

  // Intercept console.log/warn/error
  const origLog = console.log
  const origWarn = console.warn
  const origError = console.error

  console.log = (...args: unknown[]) => {
    addBreadcrumb({ category: "console", message: formatArgs(args), level: "info" })
    origLog.apply(console, args)
  }
  console.warn = (...args: unknown[]) => {
    addBreadcrumb({ category: "console", message: formatArgs(args), level: "warning" })
    origWarn.apply(console, args)
  }
  console.error = (...args: unknown[]) => {
    addBreadcrumb({ category: "console", message: formatArgs(args), level: "error" })
    origError.apply(console, args)
  }

  // Intercept fetch
  if (typeof globalThis.fetch === "function") {
    const origFetch = globalThis.fetch
    globalThis.fetch = async (input: RequestInfo | URL, init?: RequestInit) => {
      const rawUrl = typeof input === "string" ? input : input instanceof URL ? input.href : input.url
      const url = scrubUrl(rawUrl)
      const method = init?.method ?? "GET"
      addBreadcrumb({ category: "fetch", message: `${method} ${url}`, level: "info" })

      // FullTrace: inject X-IW-Session-Id when same-origin (or cross-origin opt-in).
      // injectSessionHeader returns the original init if the session is inactive,
      // so this is a zero-cost no-op for users on v0.7.x behaviour.
      const finalInit = injectSessionHeader(rawUrl, init)

      try {
        const resp = await origFetch(input, finalInit)
        if (!resp.ok) {
          addBreadcrumb({ category: "fetch", message: `${method} ${url} → ${resp.status}`, level: "warning" })
        }
        return resp
      } catch (err) {
        addBreadcrumb({ category: "fetch", message: `${method} ${url} → FAILED`, level: "error" })
        throw err
      }
    }
  }
}

function formatArgs(args: unknown[]): string {
  return args.map((a) => {
    if (typeof a === "string") return a
    try { return JSON.stringify(a) } catch { return String(a) }
  }).join(" ").slice(0, 200)
}

File: go/breadcrumbs.go
// Breadcrumb ring buffer (30 slots, FIFO) + secret scrubbing +
// optional http.Client RoundTripper wrap. Mirrors the Python / Node
// SDK behaviour: process-wide ring, bounded, thread-safe.
package capture

import (
	"fmt"
	"net/http"
	"net/url"
	"regexp"
	"sync"
	"time"
)

const maxBreadcrumbs = 30

var (
	breadcrumbMu sync.Mutex
	breadcrumbs  = make([]Breadcrumb, 0, maxBreadcrumbs)
)

var secretPatterns = []*regexp.Regexp{
	regexp.MustCompile(`Bearer\s+[A-Za-z0-9\-._~+/]+=*`),
	regexp.MustCompile(`[a-zA-Z0-9_-]{20,}\.[a-zA-Z0-9_-]{20,}\.[a-zA-Z0-9_-]{20,}`),
	regexp.MustCompile(`(?i)(?:sk|pk|api|key|token|secret|password|passwd)[_-]?[:\s=]+\S{8,}`),
	regexp.MustCompile(`://[^:/]+:[^@]+@`),
	regexp.MustCompile(`(?i)[?&](api_key|token|secret|key|password|auth|credential)=[^&\s]+`),
}

func scrubSecrets(text string) string {
	out := text
	for _, p := range secretPatterns {
		out = p.ReplaceAllString(out, "[REDACTED]")
	}
	return out
}

var sensitiveQueryParams = map[string]struct{}{
	"token": {}, "key": {}, "secret": {}, "password": {}, "auth": {},
	"credential": {}, "api_key": {}, "apiKey": {}, "access_token": {},
}

// ScrubURL replaces sensitive query parameters with [REDACTED].
func ScrubURL(raw string) string {
	u, err := url.Parse(raw)
	if err != nil {
		return scrubSecrets(raw)
	}
	if u.RawQuery == "" {
		return scrubSecrets(raw)
	}
	vals := u.Query()
	for k := range vals {
		if _, sensitive := sensitiveQueryParams[k]; sensitive {
			vals.Set(k, "[REDACTED]")
		}
	}
	u.RawQuery = vals.Encode()
	return u.String()
}

// AddBreadcrumb appends an entry to the ring. Secrets inside
// ``message`` are scrubbed via secretPatterns before storage. Messages
// longer than 200 chars are truncated.
func AddBreadcrumb(b Breadcrumb) {
	if b.Timestamp == "" {
		b.Timestamp = nowISO()
	}
	if b.Category == "" {
		b.Category = "custom"
	}
	if b.Level == "" {
		b.Level = "info"
	}
	msg := b.Message
	if len(msg) > 200 {
		msg = msg[:200]
	}
	b.Message = scrubSecrets(msg)

	breadcrumbMu.Lock()
	defer breadcrumbMu.Unlock()
	if len(breadcrumbs) >= maxBreadcrumbs {
		breadcrumbs = append(breadcrumbs[1:], b)
	} else {
		breadcrumbs = append(breadcrumbs, b)
	}
}

// GetBreadcrumbs returns a snapshot of the ring buffer.
func GetBreadcrumbs() []Breadcrumb {
	breadcrumbMu.Lock()
	defer breadcrumbMu.Unlock()
	if len(breadcrumbs) == 0 {
		return nil
	}
	out := make([]Breadcrumb, len(breadcrumbs))
	copy(out, breadcrumbs)
	return out
}

// ClearBreadcrumbs — for tests.
func ClearBreadcrumbs() {
	breadcrumbMu.Lock()
	breadcrumbs = breadcrumbs[:0]
	breadcrumbMu.Unlock()
}

func nowISO() string {
	// Node's new Date().toISOString() uses UTC + millisecond precision
	// + trailing Z. time.Time's RFC3339Nano prints nanoseconds, so we
	// format manually for cross-SDK consistency.
	t := time.Now().UTC()
	return t.Format("2006-01-02T15:04:05.000Z")
}

// ── http.Client auto-intercept ──────────────────────────────────────

type breadcrumbTransport struct {
	inner http.RoundTripper
}

// WrapHTTPClient returns a RoundTripper that records every outbound
// request + response status as a breadcrumb. Opt-in per client; we
// don't globally patch http.DefaultTransport to stay friendly to apps
// that wire their own transports.
func WrapHTTPClient(inner http.RoundTripper) http.RoundTripper {
	if inner == nil {
		inner = http.DefaultTransport
	}
	return &breadcrumbTransport{inner: inner}
}

func (t *breadcrumbTransport) RoundTrip(req *http.Request) (*http.Response, error) {
	safeURL := ScrubURL(req.URL.String())
	AddBreadcrumb(Breadcrumb{
		Category: "http",
		Level:    "info",
		Message:  req.Method + " " + safeURL,
	})
	resp, err := t.inner.RoundTrip(req)
	if err != nil {
		AddBreadcrumb(Breadcrumb{
			Category: "http",
			Level:    "error",
			Message:  fmt.Sprintf("%s %s -> FAILED", req.Method, safeURL),
		})
		return resp, err
	}
	if resp.StatusCode >= 400 {
		AddBreadcrumb(Breadcrumb{
			Category: "http",
			Level:    "warning",
			Message:  fmt.Sprintf("%s %s -> %d", req.Method, safeURL, resp.StatusCode),
		})
	}
	return resp, nil
}
